### PR TITLE
Classifier RBM (WIP)

### PR DESCRIPTION
**Pull request by AdrianLsk**
Hi everyone, I finally managed to get the Classification RBM working. Corresponding paper http://www.jmlr.org/papers/volume13/larochelle12a/larochelle12a.pdf

Currently works only for RBM with cd and pcd and both samplers (Sequential and Driven).

DBM still needs work, because the model doesn't converge for predictions and it gets stuck at low accuracy for some reason.

So far no tests, hence WIP.

Still need to experiment with non-Bernoulli inputs, Gaussian seems to work but still need to properly test it altogether with Exponential or Ising.

### CODE REVIEW

File: paysage/backends/python_backend/nonlinearity.py [hunk 1]
```diff
@@ -275,3 +275,17 @@ def sin(x: T.Tensor) -> T.Tensor:
...
+    Elementwise softplus function of a tensor.
+
+    Args:
+        x: A tensor.
+
+    Returns:
+        tensor: Elementwise softplus.
+
+    """
+    y = (x - ne.evaluate('max(x)')).T.astype('f4')
```
Comment (AdrianLsk) [inline paysage/backends/python_backend/nonlinearity.py]: agree

File: examples/example_util.py [hunk 2]
```diff
@@ -53,13 +53,18 @@ def show_metrics(rbm, performance):
 
 def compute_reconstructions(rbm, v_data, fit, n_recon=10, vertical=False):
     sampler = fit.DrivenSequentialMC(rbm)
-    data_state = State.from_visible(v_data, rbm)
+    data = v_data[0] if isinstance(v_data, tuple) else v_data
```
Comment (AdrianLsk) [inline examples/example_util.py]: the reason is line 63, where I want to know if the input contains also targets, if so, I want to pull out predictions and print them

File: paysage/backends/config.json [hunk 3]
```diff
@@ -1 +1 @@
-{"backend": "python"}
\ No newline at end of file
+{"backend": "pytorch"}
```
Comment (AdrianLsk) [inline paysage/backends/config.json]: ok

File: paysage/backends/pytorch_backend/matrix.py [hunk 4]
```diff
@@ -1378,4 +1378,33 @@ def fast_energy_distance(minibatch: T.FloatTensor,
             d3 += euclidean_distance(X[i], Y[j])
     d3 = d3 / (n*m)
 
-    return 2.0 * d3 - d2 - d1
+    return float_tensor(numpy.array([2.0 * d3 - d2 - d1]))
+
+
+def int_tensor(tensor: T.Tensor) -> T.FloatTensor:
```
Comment (AdrianLsk) [inline paysage/backends/pytorch_backend/matrix.py]: sure

File: paysage/backends/pytorch_backend/matrix.py [hunk 5]
```diff
@@ -1378,4 +1378,33 @@ def fast_energy_distance(minibatch: T.FloatTensor,
...
+    """
+    try:
+        # tensor is a numpy object
+        return torch.IntTensor(tensor.astype(int))
+    except Exception:
+        # tensor is a torch object
+        return tensor.int()
+
+
+def onehot(tensor, n_classes: int) -> T.FloatTensor:
```
Comment (AdrianLsk) [inline paysage/backends/pytorch_backend/matrix.py]: I will try to use a sparse tensor instead

Comment (AdrianLsk) [inline paysage/backends/pytorch_backend/matrix.py]: Unfortunately, the sparse tensors do not support dot products and scatter_ does not support tensors of different shapes, i.e. the output tensor is of shape (n_samples, n_classes) while the input tensor has shape (n_samples,). So for now this should be OK, I guess.

File: examples/example_util.py (hunk 2 above)
Comment (jrwalsh1) [inline examples/example_util.py]: It would be better if each function had a single purpose, and the types of the arguments are clear.  A distinct function should handle displaying class predictions.  

Separate from this PR, some of these functions are lacking docstrings.

File: paysage/fit.py [hunk 6]
```diff
@@ -510,15 +536,16 @@ def tap(vdata, model, sampler, positive_steps=1, init_lr_EMF=0.1, tolerance_EMF=
...
+                              max_iters_EMF)
 
 
 class StochasticGradientDescent(object):
     """Stochastic gradient descent with minibatches"""
     def __init__(self, model, batch, optimizer, epochs, method=pcd,
                  sampler=SequentialMC, mcsteps=1, monitor=None):
         """
-        Create a StochasticGradientDescent object.
+        Create a StochasticG([x.shape for x in units])radientDescent object.
```
Comment (jrwalsh1) [inline paysage/fit.py]: fix

File: paysage/metrics.py [hunk 7]
```diff
@@ -394,3 +394,139 @@ def value(self) -> float:
             return self.calc.var
         else:
             return None
+
+
+class CrossEntropy(object):
+    """
+    Compute the root-mean-squared error between observations and their
```
Comment (jrwalsh1) [inline paysage/metrics.py]: fix the docstrings in this file, they are all duplicates of the `ReconstructionError` class.

File: paysage/backends/python_backend/matrix.py [hunk 8]
```diff
@@ -1309,3 +1310,41 @@ def fast_energy_distance(minibatch, samples, downsample=100):
...
+
+    Args:
+        predictions: A tensor.
+        targets: A tensor.
+
+    Returns:
+        float: A cross-entropy.
+
+    """
+    return - numpy.log(numpy.clip(predictions[:, targets], 1e-7, 1))
```
Comment (jrwalsh1) [inline paysage/backends/python_backend/matrix.py]: I do not think this is doing the correct thing - targets is indexing the predictions array?

File: paysage/metrics.py [hunk 9]
```diff
@@ -394,3 +394,139 @@ def value(self) -> float:
...
+
+        Args:
+            update_args: uses visible layer of minibatch and reconstructions
+
+        Returns:
+            None
+
+        """
+        xe = be.xe(update_args.reconstructions.units[-1],
+                   be.argmax(update_args.minibatch.units[-1], axis=1))
```
Comment (jrwalsh1) [inline paysage/metrics.py]: is this correct? `argmax` is only selecting single values from each row.  Are `minibatch.units[-1]` going to be class labels?

File: paysage/models/factored.py [hunk 10]
```diff
@@ -0,0 +1,4 @@
+"""
```
Comment (jrwalsh1) [inline paysage/models/factored.py]: please drop this file for this PR

File: paysage/models/initialize/init_model.py [hunk 11]
```diff
@@ -36,7 +36,7 @@ def hinton(batch, model):
                         0.01 * be.randn(model.weights[i].shape)
     while True:
         try:
-            v_data = batch.get(mode='train')
+            v_data = batch.get(mode='train')[0]
```
Comment (jrwalsh1) [inline paysage/models/initialize/init_model.py]: should this be here? initialization should happen on the whole batch.

File: paysage/models/initialize/init_model.py [hunk 12]
```diff
@@ -77,7 +77,7 @@ def glorot_normal(batch, model):
                         sigma * be.randn(model.weights[i].shape)
     while True:
         try:
-            v_data = batch.get(mode='train')
+            v_data = batch.get(mode='train')[0]
```
Comment (jrwalsh1) [inline paysage/models/initialize/init_model.py]: same as L39

File: paysage/models/state_.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
+# coding: utf-8
```
Comment (jrwalsh1) [inline paysage/models/state_.py]: This file shouldn't be needed.  We're only intending to support python3 in `paysage`.

File: paysage/backends/python_backend/matrix.py (hunk 8 above)
Comment (AdrianLsk) [inline paysage/backends/python_backend/matrix.py]: Predictions here are softmax outputs. When you calculate cross-entropy, you multiply onehot-encoded targets with logarithm of predictions and sum over the last dimension - number of classes. If the encoded target has 1 on the jth index, only jth class prediction will be taken into account and the others will be multiplied by zero. This is equivalent to selecting only the jth prediction and apply logarithm, therefore no summing with the zero-ed predictions.

pytorch is doing the same thing (but applied to logits instead of softmax) http://pytorch.org/docs/master/nn.html#torch.nn.CrossEntropyLoss

File: paysage/metrics.py (hunk 9 above)
Comment (AdrianLsk) [inline paysage/metrics.py]: Targets in `minibatch.units[-1]` are always onehot encoded with the preprocessor included in the batch object. This is necessary for the algorithm described in http://www.jmlr.org/papers/volume13/larochelle12a/larochelle12a.pdf.

Since my implementation of cross entropy - `be.xe` follows the one from pytorch http://pytorch.org/docs/master/nn.html#torch.nn.functional.cross_entropy, it accepts only the class labels instead of their onehot encoded versions. Hence the argmax over the class dimension.

**Comment by AdrianLsk** [comment]
> My main concern with the changes proposed is that they're trying to fit in the class targets object into objects like State, where they may not belong.

Data represents the visible units. That means only inputs in case of unsupervised learning and a pair of inputs and targets in the supervised case, as is a standard in ML literature.

Since the `State` object is a container for these, it should hold targets too, because the classification RBM algorithm samples them also during training as described in http://www.jmlr.org/papers/volume13/larochelle12a/larochelle12a.pdf.

> Things like if isinstance(data, tuple): make the code much harder to understand, as the type of data is no longer clear and is context dependent. Simplicity is critical as the execution stack can be deep.

I agree that simplicity is critical, but if you want to include also a classification feature, understanding the context, i.e. difference between unsupervised and supervised case will be necessary. I understand your concern, but if you look at it also in a supervised way, it makes perfect sense, since the data is a pair of inputs and targets, comprising visible units, both on opposite sides of the RBM graph.

> It may be helpful to give a qualitative overview of how you want to integrate a classification layer into the code and what the code needs to accommodate.

I tried to make as little changes as possible to preserve most of your code unchanged. The code should accommodate the possibility of taking also targets as another type of input that comprises visible units. 

And that has to start with the `Batch` object. The only way I came up with was to make the batch yield data as a single input or a pair of inputs and targets. From there I had to adjust the `State` object to handle storing the targets. Then the alternating updates are straightforward and just a matter of handling another layer, in terms of updating weights.

The classification layer can be renamed to softmax/categorical layer, since you can use also Gaussian layer for regression or Bernoullli layer for binary classification.

The part that makes it work (and learn) is that the code can handle the targets on the opposite side of the RBM graph. This means that the weights are updated by the correct errors, which are produced by the visible target values and the sampled target values. For the visible ones, clamping the last layer as a target layer is the most IMPORTANT part. This works only for RBM and unfortunately not for DBM, which needs more investigation.

File: examples/example_util.py (hunk 2 above)
Comment (AdrianLsk) [inline examples/example_util.py]: The whole point here is that if you sample image reconstructions of the input images using the `deterministic_iteration`, then you also sample their target predictions. So I thought that it would be nice to see that too.

But if you find it as a main bottleneck here, I will remove it. For me, the target classification feature is more important than this visualization feature.

**Comment by jrwalsh1** [comment]
Hi @AdrianLsk,

I'm closing this for now as it still needs a lot of work - e.g. tests, removing unnecessary components, and an example usage.  There are 2 reasons:

1. This doesn't meet the [contribution guidelines](https://github.com/drckf/paysage/blob/master/CONTRIBUTING.md), which are designed to maintain the quality and direction of the codebase.

2. As you can see, we've updated the repo with a lot of improvements, and will require some restructuring of your code.

Apologies that this update makes merging your PR more difficult. Hopefully the improvements from the recent updates will help.  I'd suggest submitting changes in smaller pieces that are easier to merge in, as well as opening an issue discussing the direction you'd like to take.
